Core/Services/StitchCoreRemoteMongoDBService/Sources/StitchCoreRemoteMongoDBService/Sync/DocumentVersionInfo.swift

```diff
@@ -0,0 +1,159 @@
+import MongoSwift
+
+internal let DOCUMENT_VERSION_FIELD = "__stitch_sync_version"
+
+/**
+ Returns a query filter for the given document _id and version. The version is allowed to be
+ null. The query will match only if there is either no version on the document in the database
+ in question if we have no reference of the version or if the version matches the database's
+ version.
+
+ - parameter documentId: the _id of the document.
+ - parameter version: the expected version of the document, if any.
+ - returns a query filter for the given document _id and version for a remote operation.
+ */
+private func getVersionedFilter(documentId: BSONValue?,
+                                version: BSONValue?) -> Document {
+    var filter: Document = ["_id": documentId]
+    if version == nil {
+        filter[DOCUMENT_VERSION_FIELD] = ["$exists": false] as Document
+    } else {
+        filter[DOCUMENT_VERSION_FIELD] = version
+    }
+    return filter
+}
+
+final class DocumentVersionInfo {
+    let version: Version?
+    let versionDoc: Document?
+    /**
+     Gets a filter that will only return the document in a query if
+     it matches the current version. Will return null if a document
+     id was not specified when the version info was constructed.
+     */
+    let filter: Document?
+    /**
+     Given a DocumentVersionInfo, returns a BSON document representing the next version. This means
+     and incremented version count for a non-empty version, or a fresh version document for
+     empty version.
+     */
+    lazy var nextVersion: Document = {
+        guard hasVersion,
+            let versionDoc = versionDoc,
+            let version = version else {
```

Comment:
[minor] isn't `hasVersion` redundant with `let version = version` based on the definition of hasVersion below?

Reply:
Good point, but that raises other questions of how this is setup.

Reply:
which are..? :)

Reply:
Questions that I stated we should kick downfield until the versioning ticket. Different semantics means there's a conversation worth being had about the original implementation.